criu/files-reg.c

```diff
@@ -1380,12 +1381,299 @@ static bool should_check_size(int flags)
+
+	note_header_end = (Elf64_Nhdr *) (file_header_end - sizeof(Elf64_Nhdr));
+	/* Arbitrary large number to prevent unnecessarily searching the entire file */
+	num_iterations = 500;
+
+	/* The note type for the build-id is NT_GNU_BUILD_ID. */
+	while (num_iterations-- && note_header <= note_header_end &&
+			note_header->n_type != NT_GNU_BUILD_ID)
+		note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+						note_header->n_namesz + note_header->n_descsz);
+
+	if (!num_iterations || note_header >= note_header_end)
+		goto err;
+
+	/* 
+	 * If the size of the notes description is too large or is invalid
+	 * then the build-id could not be obtained.
+	 */
+	if (note_header->n_descsz <= 0 || note_header->n_descsz > 512)
+		goto err;
+
+	size = note_header->n_descsz;
+	note_header = (Elf64_Nhdr *) ((char *) note_header + sizeof(Elf64_Nhdr) +
+					note_header->n_namesz);
+	note_header_end = (Elf64_Nhdr *) (file_header_end - size);
+	if (note_header <= (Elf64_Nhdr *) file_header || note_header > note_header_end)
+		goto err;
+
+	*build_id = (unsigned char *) xmalloc(size);
+	if (!*build_id)
+		goto err;
+
+	memcpy(*build_id, (void *) note_header, size);
+
+	munmap(file_header, min_t(size_t, fd_status->st_size, 1048576));
+	return size;
+	
+err:
+	munmap(file_header, min_t(size_t, fd_status->st_size, 1048576));
+	return -1;
+}
+
+/*
+ * Finds the build-id of the file by checking if the file is an ELF file
+ * and then calling either the 32-bit or the 64-bit function as necessary.
+ * Returns the number of bytes of the build-id if it could be
+ * obtained, else -1.
+ */
+static int get_build_id(const int fd, const struct stat *fd_status,
+				unsigned char **build_id)
+{
+	char *start_addr, buf[5];
+
+	if (read(fd, buf, 5) != 5)
+		return -1;
+
+	/* 
+	 * The first 4 bytes contain a magic number identifying the file as an
+	 * ELF file. They contain the characters ‘\x7f’, ‘E’, ‘L’, and
+	 * ‘F’, respectively.
+	 */
+	if (buf[0] != 0x7f || buf[1] != 0x45 || buf[2] != 0x4c || buf[3] != 0x46)
+		return -1;
+
+	/* 
+	 * If the build-id exists, then it will most likely be present in the
+	 * beginning of the file. Therefore at most only the first 1 MB of the 
+	 * file is mapped.
+	 */
+	start_addr = (char *) mmap(0, min_t(size_t, fd_status->st_size, 1048576),
+					PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (start_addr == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return -1;
+	}
+
+	if (buf[4] == ELFCLASS32)
+		return get_build_id_32((Elf32_Ehdr *) start_addr, fd_status, build_id);
+	if (buf[4] == ELFCLASS64)
+		return get_build_id_64((Elf64_Ehdr *) start_addr, fd_status, build_id);
+	
+	munmap(start_addr, min_t(size_t, fd_status->st_size, 1048576));
+	return -1;
+}
+
+/*
+ * Finds and stores the build-id of a file, if it exists, so that it can be validated
+ * while restoring.
+ * Returns 1 if the build-id of the file could be stored, 0 if there was an error
+ * or -1 if the build-id could not be obtained.
```

Comment:
I think it would be more consistent if it will return -1 on an error and 0 if the build-id could not be obtained.

Reply:
If I do this, I won't be able to simply return !!result in store_validation_data().

Reply:
Then I won't be able to simply return !!result in store_validation_data().

Reply:
You will need to write three lines of code and it will be more readable:
if (ret < 0)
  return -1;
return 0;

Reply:
I've pushed this change